opm/output/eclipse/EclipseReader.cpp

```diff
@@ -17,230 +17,163 @@
+    inline data::Solution restoreSOLUTION( ecl_file_type* file,
+                                           int numcells,
+                                           const UnitSystem& units ) {
 
-            float* pressure_data = ecl_kw_get_float_ptr(pressure_kw);
-            const double deck_pressure_unit = (eclipse_state->getDeckUnitSystem().getType() == UnitSystem::UNIT_TYPE_METRIC) ? Opm::unit::barsa : Opm::unit::psia;
-            for (size_t index = 0; index < simulator_state.pressure().size(); ++index) {
-                simulator_state.pressure()[index] = unit::convert::from((double)pressure_data[index], deck_pressure_unit);
-            }
-        } else {
-            throw std::runtime_error("Read of restart file: File does not contain PRESSURE data\n");
-        }
-    }
-
-
-    static void restoreSaturation(const ecl_file_type* file_type,
-                                  const PhaseUsage& phaseUsage,
-                                  int numcells,
-                                  SimulationDataContainer& simulator_state) {
-
-        float* sgas_data = NULL;
-        float* swat_data = NULL;
-
-        if (phaseUsage.phase_used[BlackoilPhases::Aqua]) {
-            const char* swat = "SWAT";
-            if (ecl_file_has_kw(file_type, swat)) {
-                ecl_kw_type* swat_kw = ecl_file_iget_named_kw(file_type , swat, 0);
-                swat_data = ecl_kw_get_float_ptr(swat_kw);
-                std::vector<double> swat_datavec(&swat_data[0], &swat_data[numcells]);
-                EclipseIOUtil::addToStripedData(swat_datavec, simulator_state.saturation(), phaseUsage.phase_pos[BlackoilPhases::Aqua], phaseUsage.num_phases);
-            } else {
-                std::string error_str = "Restart file is missing SWAT data!\n";
-                throw std::runtime_error(error_str);
-            }
+        for( const auto* key : { "PRESSURE", "TEMP", "SWAT", "SGAS" } ) {
+            if( !ecl_file_has_kw( file, key ) )
+                throw std::runtime_error("Read of restart file: "
+                                         "File does not contain "
+                                         + std::string( key )
+                                         + " data" );
         }
 
-        if (phaseUsage.phase_used[BlackoilPhases::Vapour]) {
-            const char* sgas = "SGAS";
-            if (ecl_file_has_kw(file_type, sgas)) {
-                ecl_kw_type* sgas_kw = ecl_file_iget_named_kw(file_type , sgas, 0);
-                sgas_data = ecl_kw_get_float_ptr(sgas_kw);
-                std::vector<double> sgas_datavec(&sgas_data[0], &sgas_data[numcells]);
-                EclipseIOUtil::addToStripedData(sgas_datavec, simulator_state.saturation(), phaseUsage.phase_pos[BlackoilPhases::Vapour], phaseUsage.num_phases);
-            } else {
-                std::string error_str = "Restart file is missing SGAS data!\n";
-                throw std::runtime_error(error_str);
-            }
+        struct keyword {
```

Comment:
Hopefully you could benefit from using: https://github.com/Ensembles/ert/blob/master/devel/libeclxx/include/ert/ecl/EclKW.hpp